websearch/websearch/src/durability.rs

```diff
@@ -0,0 +1,579 @@
+            }
+        }
+
+        #[allow(dead_code)]
+        fn subscribe(&self) -> Pollable {
+            let mut state = self.state.borrow_mut();
+            match &mut *state {
+                Some(DurableSearchSessionState::Live { session, .. }) => Impl::subscribe(session),
+                Some(DurableSearchSessionState::Replay { pollables, .. }) => {
+                    let lazy_pollable = LazyInitializedPollable::new();
+                    let pollable = lazy_pollable.subscribe();
+                    pollables.push(lazy_pollable);
+                    pollable
+                }
+                None => {
+                    unreachable!()
+                }
+            }
+        }
+    }
+
+    impl<Impl: ExtendedwebsearchGuest> Drop for DurableSearchSession<Impl> {
+        fn drop(&mut self) {
+            let _ = self.subscription.take();
+            match self.state.take() {
+                Some(DurableSearchSessionState::Live {
+                    mut pollables,
+                    session,
+                }) => {
+                    with_persistence_level(PersistenceLevel::PersistNothing, move || {
+                        pollables.clear();
+                        drop(session);
+                    });
+                }
+                Some(DurableSearchSessionState::Replay { mut pollables, .. }) => {
+                    pollables.clear();
+                }
+                None => {}
+            }
+        }
+    }
+
+    impl<Impl: ExtendedwebsearchGuest> GuestSearchSession for DurableSearchSession<Impl> {
+        fn next_page(&self) -> Result<Vec<SearchResult>, SearchError> {
+            let durability = Durability::<Vec<SearchResult>, UnusedError>::new(
+                "golem_websearch",
+                "next_page",
+                DurableFunctionType::ReadRemote,
+            );
+            if durability.is_live() {
+                let mut state = self.state.borrow_mut();
+                let (result, new_live_session) = match &*state {
+                    Some(DurableSearchSessionState::Live { session, .. }) => {
+                        let result =
+                            with_persistence_level(PersistenceLevel::PersistNothing, || {
+                                session.next_page()
+                            });
+                        let value = result?;
+                        (Ok(durability.persist_infallible(NoInput, value)), None)
+                    }
+                    Some(DurableSearchSessionState::Replay {
+                        original_params,
+                        pollables,
+                        partial_results,
+                        finished,
+                        ..
+                    }) => {
+                        if *finished {
+                            (Ok(Vec::new()), None)
+                        } else {
+                            let retry_params = Impl::retry_params(original_params, partial_results);
```

Comment:
As partial results are never written to, I don't believe that this will currently produce the correct parameters for the next page.